### PR TITLE
combine subspecies

### PR DESCRIPTION
**Pull request by davharris**
like #40, but with cleaner history

### CODE REVIEW

**Comment by ethanwhite** [comment]
@sdtaylor - can you give this a review when you get a few minutes.

**Comment by sdtaylor** [comment]
does the package RPostgres work for both of you? The only one I have via CRAN is RPostgreSQL, which looks like it's exactly the same thing but with a different name.

**Comment by davharris** [comment]
I just used the one that was used in `filter_species`.  It works fine for me, except it doesn’t seem to respect my `PGUSER` variable.

**Comment by sdtaylor** [comment]
working around it for now. It shouldn't be an issue once everything is switched to sqlite in #38

**Comment by ethanwhite** [comment]
https://github.com/rstats-db/RPostgres

File: forecast-bbs-core.R [hunk 1]
```diff
@@ -35,28 +35,78 @@ install_dataset <- function(dataset){
...
+  # Add the abundances for subspecies that co-occurred at the same site in the
+  # same year. Takes about 3 minutes on my laptop.
+  combine = function(x){
+    x$abundance = sum(x$abundance)
+    x[1, ]
+  }
+  fixed_subspecies = df %>% 
+    filter(species_id %in% new_subspecies_ids) %>%
+    group_by(site_id, year, species_id) %>%
+    do(combine(.)) %>%
```
Comment (sdtaylor) [inline forecast-bbs-core.R]: This part takes 5-6 minutes on my laptop. But replacing `do(combine(.))` with `summarize(abundance = sum(abundance))` takes only a few seconds and gives the same answer. What's reason for the former?

Comment (davharris) [inline forecast-bbs-core.R]: Um.  Don't think there was actually a good reason, except that `summarize()` drops columns. But we could add those back in easily.  I'll do that now.

Thanks for pointing this out.

**Comment by davharris** [comment]
@sdtaylor would you mind reviewing this version? Thanks for the dplyr::summarize suggestion.

File: forecast-bbs-core.R [hunk 2]
```diff
@@ -28,35 +28,75 @@ install_dataset <- function(dataset){
...
-    filter(species_id < 3900 | species_id > 3910) %>%
-    filter(species_id < 4160 | species_id > 4210) %>%
-    filter(species_id != 7010) %>%
-    filter(species_id %in% identified_taxa)
+    filter(species_id %in% new_subspecies_ids) %>%
+    group_by(site_id, year, species_id) %>%
+    summarize(abundance = sum(abundance)) %>%
+    left_join(dplyr::select(df, -abundance),
+              by = c("site_id", "year", "species_id"))
+}
```
Comment (sdtaylor) [inline forecast-bbs-core.R]: without the bind_rows() call this is now returning only those subspecies that needed adjusting. I think this is what you want. it works for me and preserves the lat long columns. 

'''
   fixed_subspecies = df %>% 
    filter(species_id %in% new_subspecies_ids) %>%
    group_by(site_id, year, species_id, lat, long) %>%
    summarize(abundance = sum(abundance)) %>%
    ungroup()
  
  # bind the rows of the combined subspecies with the rows of everyone else
 df %>%
    filter(!(species_id %in% new_subspecies_ids)) %>%
    bind_rows(fixed_subspecies)
'''

Comment (davharris) [inline forecast-bbs-core.R]: Yikes, good catch.  Couldn't I also just omit `filter(species_id %in% identified_taxa)`? That's what I had intended to do.

Thanks again for the helpful comments on this.

Comment (davharris) [inline forecast-bbs-core.R]: Wait, nevermind.  Rereading....

Comment (davharris) [inline forecast-bbs-core.R]: Sorry for the triple comment.  I meant to say, "omit `filter(species_id %in% new_subspecies_ids)`"  That way, we'd never collapse down to just the new species.  Now that we're using summarize, which is fast, I don't see a reason to split out the subspecies anymore.

Comment (sdtaylor) [inline forecast-bbs-core.R]: ok that makes more sense. But now that I run it, the total rows in df isn't being reduced like it should be from aggregating species. It looks like the left join is adding things back in somehow. I think we can just get rid of the join and put  lat, long in the `group_by()` part to retain them. lat longs are unique to each site_id, so it's not adding "extra groups". 

df %>%
    group_by(site_id, year, species_id, lat, long) %>%
    summarize(abundance = sum(abundance)) %>%
    ungroup()

**Comment by davharris** [comment]
Thanks @sdtaylor! I changed those lines like you suggested.  Ready to merge?

**Comment by sdtaylor** [comment]
Yep, looks good.

**Pull request merged**